tfmesos/scheduler.py

```diff
@@ -359,23 +372,55 @@ def registered(self, driver, framework_id, master_info):
             )
 
     def statusUpdate(self, driver, update):
-        mesos_task_id = int(update.task_id.value)
-        if update.state != 'TASK_RUNNING':
+        logger.debug('Received status update %s', str(update.state))
+        mesos_task_id = update.task_id.value
+        if self._is_terminal_state(update.state):
             task = self.tasks[mesos_task_id]
             if self.started:
                 if update.state != 'TASK_FINISHED':
-                    logger.error('Task failed: %s, %s', task, update.message)
+                    logger.error('Task failed: %s, %s with state %s', task,
+                                 update.message, update.state)
                     raise RuntimeError(
-                        'Task %s failed! %s' % (task, update.message)
+                        'Task %s failed! %s with state %s' %
+                        (task, update.message, update.state)
                     )
                 else:
                     self.job_finished[task.job_name] += 1
             else:
-                logger.warn('Task failed: %s, %s', task, update.message)
+                logger.warn('Task failed while launching the server: %s, '
+                            '%s with state %s', task,
+                            update.message, update.state)
+
                 if task.connection:
                     task.connection.close()
 
-                driver.reviveOffers()
+                self.task_failure_count[self.decorated_task_index(task)] += 1
+
+                if self.can_revive_task(task):
+                    logger.info('Going to revive task %s ', task.task_index)
+                    self.tasks.pop(mesos_task_id)
+                    task.offered = False
```

Comment:
Because the task has failed, those fields would have never got set in the first place. But agree that it is better to reset these values